python/ray/_private/gcs_pubsub.py

```diff
@@ -94,108 +122,181 @@ def publish_error(self, key_id: bytes, error_info: ErrorTableData) -> None:
-        self._errors = deque()
-        # Future for indicating whether the subscriber has closed.
+
+        # Type of the channel.
+        self._channel = pubsub_channel_type
+        # Protects multi-threaded read and write of self._queue.
+        self._lock = threading.Lock()
+        # A queue of received PubMessage.
+        self._queue = deque()
+        # Indicates whether the subscriber has closed.
         self._close = threading.Event()
 
-    def subscribe_error(self) -> None:
-        """Registers a subscription for error info.
+    def subscribe(self) -> None:
+        """Registers a subscription for the subscriber's channel type.
 
-        Before the registration, published errors will not be saved for the
-        subscriber.
+        Before the registration, published messages in the channel will not be
+        saved for the subscriber.
         """
         with self._lock:
             if self._close.is_set():
                 return
-            if not self._subscribed_error:
-                req = self._subscribe_error_request()
-                self._stub.GcsSubscriberCommandBatch(req, timeout=30)
-                self._subscribed_error = True
+            req = self._subscribe_request(self._channel)
+            self._stub.GcsSubscriberCommandBatch(req, timeout=30)
 
-    def poll_error(self, timeout=None) -> Tuple[bytes, ErrorTableData]:
-        """Polls for new error messages."""
-        with self._lock:
+    def _poll(self, timeout=None) -> None:
+        assert self._lock.locked()
+
+        # Poll until data becomes available.
+        while len(self._queue) == 0:
             if self._close.is_set():
                 return
 
-            if len(self._errors) == 0:
-                req = self._poll_request()
-                fut = self._stub.GcsSubscriberPoll.future(req, timeout=timeout)
-                # Wait for result to become available, or cancel if the
-                # subscriber has closed.
-                while True:
-                    try:
-                        fut.result(timeout=1)
-                        break
-                    except grpc.FutureTimeoutError:
-                        # Subscriber has closed. Cancel inflight the request
-                        # and return from polling.
-                        if self._close.is_set():
-                            fut.cancel()
-                            return None, None
-                        # GRPC has not replied, continue waiting.
+            fut = self._stub.GcsSubscriberPoll.future(
+                self._poll_request(), timeout=timeout)
+            # Wait for result to become available, or cancel if the
+            # subscriber has closed.
+            while True:
+                # Unlock before waiting for reply.
+                self._lock.release()
+                try:
+                    # Use 1s timeout to check for subscriber closing
+                    # periodically.
+                    fut.result(timeout=1)
```

Comment:
Do we still need to worry about async closes if everything is sync? Seems like we can remove this logic and also just hold the lock during the entire poll, removing the need for the loop at all.

Reply:
Unfortunately we still need the inner loop to check for async closes. Currently in workers, polling errors and logs are done in threads different from where subscriber.close() is called. But I simplified locking because we no longer need to lock in close().